Improve error when there's an unrecognized message from the server

Namely https://github.com/andreykaipov/goobs/blob/v0.7.1/client.go#L206:

```
panic: idk what kinda message this is lol #19 
```
 
See https://github.com/muesli/obs-cli/issues/19. In that case, when there's a client/server version discrepancy.

**Ideally we shouldn't panic at all.** Not quite sure if the flow makes sense yet, but maybe I can just wrap those errors as events like I do if any other errors in that loop. The client would then properly error when getting the auth required (if any).

--- a/api/requests/client.go
+++ b/api/requests/client.go
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	"time"
 
 	"github.com/gorilla/websocket"
 	uuid "github.com/nu7hatch/gouuid"
@@ -22,6 +23,9 @@ type Logger interface {
 type Client struct {
 	// The backing websocket connection to the OBS websocket server.
 	Conn *websocket.Conn
+
+	// The time we're willing to wait to receive a response from the server.
+	ResponseTimeout time.Duration
 
 	// Raw JSON message responses from the websocker server.
 	IncomingResponses chan json.RawMessage
@@ -85,7 +89,12 @@ func (c *Client) SendRequest(params Params, response Response) error {
 		return err
 	}
 
-	msg := <-c.IncomingResponses
+	var msg json.RawMessage
+	select {
+	case msg = <-c.IncomingResponses:
+	case <-time.After(c.ResponseTimeout * time.Millisecond):
+		return fmt.Errorf("Timed out receiving response from server for request %q", params.GetRequestType())
+	}
 
 	// OBS websocket seems to return already pretty-printed JSON in its
 	// responses, so we compact the response to print it on just one line.

--- a/api/requests/client.go
+++ b/api/requests/client.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/andreykaipov/goobs/api/events"
 	"github.com/gorilla/websocket"
 	uuid "github.com/nu7hatch/gouuid"
 )
@@ -27,7 +28,10 @@ type Client struct {
 	// The time we're willing to wait to receive a response from the server.
 	ResponseTimeout time.Duration
 
-	// Raw JSON message responses from the websocker server.
+	// Events broadcast by the server when actions happen within OBS.
+	IncomingEvents chan events.Event
+
+	// Raw JSON message responses from the websocket server.
 	IncomingResponses chan json.RawMessage
 
 	Log Logger
@@ -39,6 +43,9 @@ open connection. You don't really have to do this as any connections should
 close when your program terminates or interrupts. But here's a function anyways.
 */
 func (c *Client) Disconnect() error {
+	close(c.IncomingResponses)
+	close(c.IncomingEvents)
+
 	return c.Conn.WriteMessage(
 		websocket.CloseMessage,
 		websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""),

--- a/api/requests/requests.go
+++ b/api/requests/requests.go
@@ -9,7 +9,7 @@ type Params interface {
 	SetMessageID(string)
 
 	// The name of the actual request, i.e. "Abc" for "AbcParams". Used to
-	// set the RequestType.
+	// set the RequestType, so it's essentially an alias for GetRequestType.
 	GetSelfName() string
 }
 

--- a/client.go
+++ b/client.go
@@ -21,9 +21,7 @@ var version = "0.8.0-dev"
 
 // Client represents a client to an OBS websockets server.
 type Client struct {
-	IncomingEvents chan events.Event
 	*requests.Client
-
 	subclients
 	host          string
 	password      string
@@ -236,6 +234,12 @@ func (c *Client) handleConnection(messages chan json.RawMessage, errors chan err
 	}
 }
 
+// Handles messages from the server. They might be response bodies associated
+// with requests, or they can be events we can subscribe to via the
+// `client.IncomingEvents` channel. Or they can be something totally else, in
+// which case we expose the errors as more events! Despite also handling
+// incoming responses, we refer to this loop as the "eventing loop" elsewhere in
+// the comments.
 func (c *Client) handleRawMessages(messages chan json.RawMessage, errors chan error) {
 	for raw := range messages {
 		// Parse into a generic map to figure out if it's an event or

--- a/client.go
+++ b/client.go
@@ -245,7 +245,7 @@ func (c *Client) handleRawMessages(messages chan json.RawMessage, errors chan er
 			continue
 		}
 
-		panic("idk what kinda message this is lol")
+		errors <- fmt.Errorf("Client/server version mismatch? Unrecognized message: %s", raw)
 	}
 }
 

--- a/client.go
+++ b/client.go
@@ -199,7 +199,7 @@ func (c *Client) handleMessages() {
 // Expose eventing errors as... more events
 func (c *Client) handleErrors(errors chan error) {
 	for err := range errors {
-		c.IncomingEvents <- events.WrapError(err)
+		c.writeEvent(events.WrapError(err))
 	}
 }
 
@@ -225,7 +225,7 @@ func (c *Client) handleRawMessages(messages chan json.RawMessage, errors chan er
 			continue
 		}
 
-		// Responses are parsed when the request is sent
+		// Responses are parsed in the embedded Client's `SendRequest`
 		if _, ok := checked["message-id"]; ok {
 			c.IncomingResponses <- raw
 			continue
@@ -241,14 +241,30 @@ func (c *Client) handleRawMessages(messages chan json.RawMessage, errors chan er
 				continue
 			}
 
-			select {
-			case c.IncomingEvents <- event:
-			default:
-			}
-
+			c.writeEvent(event)
 			continue
 		}
 
 		panic("idk what kinda message this is lol")
+	}
+}
+
+// Since our events channel is buffered and might not necessarily be used, we
+// purge old events and write latest ones so that whenever somebody might want
+// to use it, they'll have the latest events available to them.
+func (c *Client) writeEvent(event events.Event) {
+	select {
+	case c.IncomingEvents <- event:
+	default:
+		if len(c.IncomingEvents) == cap(c.IncomingEvents) {
+			// incoming events was full (but might not be by now),
+			// so safely read off the oldest, and write the latest
+			select {
+			case _ = <-c.IncomingEvents:
+			default:
+			}
+
+			c.IncomingEvents <- event
+		}
 	}
 }

--- a/client.go
+++ b/client.go
@@ -9,6 +9,7 @@ import (
 	"net/http"
 	"net/url"
 	"os"
+	"time"
 
 	"github.com/andreykaipov/goobs/api/events"
 	"github.com/andreykaipov/goobs/api/requests"
@@ -77,6 +78,15 @@ func WithRequestHeader(x http.Header) Option {
 	}
 }
 
+// WithResponseTimeout sets the time we're willing to wait to receive a response
+// from the server for any request, before responding with an error. It's in
+// milliseconds. The default timeout is 10 seconds.
+func WithResponseTimeout(x time.Duration) Option {
+	return func(o *Client) {
+		o.ResponseTimeout = time.Duration(x)
+	}
+}
+
 type discard struct{}
 
 func (o *discard) Printf(format string, v ...interface{}) {}
@@ -93,8 +103,10 @@ It also opens up a connection, so be sure to check the error.
 */
 func New(host string, opts ...Option) (*Client, error) {
 	c := &Client{
-		Client: &requests.Client{},
-		host:   host,
+		Client: &requests.Client{
+			ResponseTimeout: 10000,
+		},
+		host: host,
 	}
 
 	for _, opt := range opts {

--- a/client.go
+++ b/client.go
@@ -142,8 +142,8 @@ func New(host string, opts ...Option) (*Client, error) {
 	c.IncomingResponses = make(chan json.RawMessage, 100)
 	go c.handleMessages()
 
-	if err := c.authenticate(); err != nil {
-		return nil, err
+	if err := c.wrappedAuthentication(); err != nil {
+		return nil, fmt.Errorf("Failed auth: %s", err)
 	}
 
 	return c, nil
@@ -161,12 +161,37 @@ func (c *Client) connect() (err error) {
 	return nil
 }
 
+// Handling authentication errors is a tad tricky. Because the auth request we
+// send depends on the eventing loop too, we need a way to return any errors
+// that might come up when parsing the auth response, while also handling
+// expected auth errors like bad creds.
+func (c *Client) wrappedAuthentication() error {
+	go func() {
+		if err := c.authenticate(); err != nil {
+			c.IncomingEvents <- events.WrapError(err)
+		}
+		c.IncomingEvents <- nil
+	}()
+
+	switch e := (<-c.IncomingEvents).(type) {
+	case *events.Error:
+		// this error can be from the above `authenticate()`, or from
+		// any errors that might've come up during the eventing loop
+		return e.Err
+	case nil:
+		return nil
+	default:
+		// only events as of now should be errors or our above nil
+		return fmt.Errorf("Surely impossible? How did the server send actual events before authentication?")
+	}
+}
+
 // Pretty much the pseudo-code from
 // https://github.com/Palakis/obs-websocket/blob/4.x-current/docs/generated/protocol.md#authentication
 func (c *Client) authenticate() error {
 	authReqResp, err := c.General.GetAuthRequired()
 	if err != nil {
-		return fmt.Errorf("Failed getting auth: %s", err)
+		return fmt.Errorf("Failed getting auth required: %s", err)
 	}
 
 	if !authReqResp.AuthRequired {
@@ -179,13 +204,9 @@ func (c *Client) authenticate() error {
 	authHash := sha256.Sum256([]byte(secret + authReqResp.Challenge))
 	authSecret := base64.StdEncoding.EncodeToString(authHash[:])
 
-	if _, err := c.General.Authenticate(&general.AuthenticateParams{
-		Auth: authSecret,
-	}); err != nil {
-		return fmt.Errorf("Failed authenticating: %s", err)
-	}
+	_, err = c.General.Authenticate(&general.AuthenticateParams{Auth: authSecret})
 
-	return nil
+	return err
 }
 
 func (c *Client) handleMessages() {

